onnxruntime/contrib_ops/cpu/ngram.cc

```diff
@@ -0,0 +1,512 @@
+ private:
+  std::vector<std::reference_wrapper<const std::string>> items_;
+  size_t hash_ = 0;
+
+  void recompute(const std::string& s) {
+    std::hash<std::string> hf{};
+    hash_ ^= hf(s) + 0x9e3779b9 + (hash_ << 6) + (hash_ >> 2);
+  }
+
+ public:
+  template <typename ForwardIter>
+  explicit NGramItem(size_t id, ForwardIter first, ForwardIter last) : NgramElementBase(id) {
+    while (first != last) {
+      recompute(*first);
+      items_.push_back(std::cref(*first));
+      ++first;
+    }
+    assert(!items_.empty());
+  }
+  explicit NGramItem() : NgramElementBase(0) {}
+  void AddItem(const std::string& s) {
+    items_.push_back(std::cref(s));
+    recompute(s);
+  }
+  void DebugPrint() const {
+    std::copy(items_.cbegin(), items_.cend(), std::ostream_iterator<std::string>(std::cout, ","));
+    std::cout << std::endl;
+  }
+  void Clear() {
+    items_.clear();
+    hash_ = 0;
+  }
+
+  bool operator==(const NGramItem& o) const {
+    if (items_.size() == o.items_.size()) {
+      return std::equal(items_.cbegin(), items_.cend(),
+                        o.items_.cbegin(), o.items_.cend(),
+                        std::equal_to<std::string>());
+    }
+    return false;
+  }
+  size_t hash() const {
+    return hash_;
+  }
+};
+
+using IntegerPoolSet = std::unordered_set<NGramItem<int64_t>>;
+// Does not own strings, contains references to them. This helps
+// to search by string references that point to the current input.
+using StringPoolSet = std::unordered_set<NGramItem<std::string>>;
+
+template <typename ForwardIter, typename Cont>
+inline void Emplace(ForwardIter first, size_t ngrams, size_t ngram_size, size_t& ngram_id, Cont& c) {
+  for (; ngrams > 0; --ngrams) {
+    c.emplace(ngram_id, first, first + ngram_size);
+    first += ngram_size;
+    ++ngram_id;
+  }
+}
+
+}  // namespace ngram_details
+}  // namespace contrib
+}  // namespace onnxruntime
+
+using namespace onnxruntime::contrib::ngram_details;
+
+namespace std {
+template <typename T>
+struct hash<NGramItem<T>> {
+  typedef NGramItem<T> argument_type;
+  typedef size_t result_type;
+  result_type operator()(const argument_type& a) const {
+    return a.hash();
+  }
+};
+}  // namespace std
+
+namespace onnxruntime {
+namespace contrib {
+
+enum Mode {
+  kNone = 0,
+  kTF = 1,
+  kIDF = 2,
+  kTFIDF = 3
+};
+
+struct Ngram::Impl {
+  Mode mode_ = kNone;
+  int64_t N_ = 0;
+  int64_t M_ = 0;
+  int64_t S_ = 0;
+  bool all_ = false;
+  std::vector<int64_t> ngram_counts_;
+  std::vector<int64_t> ngram_indexes_;
+  std::vector<float> weights_;
+
+  std::vector<std::string> pool_strings_;
+  StringPoolSet str_set_;
+  IntegerPoolSet int_set_;
+  size_t output_size_ = 0;
+
+  MLDataType int32_dt_;
+  MLDataType int64_dt_;
+  MLDataType string_dt_;
+  Impl() {
+    int32_dt_ = DataTypeImpl::GetType<int32_t>();
+    int64_dt_ = DataTypeImpl::GetType<int64_t>();
+    string_dt_ = DataTypeImpl::GetType<std::string>();
+  }
```

Comment:
Could these be static constants instead?

Reply:
I think we generally avoid statics or if we do, we make them function locals so they are constructed when they are needed. I'd not make them static. Generally, people do not even cache them.

---
In reply to: [242803456](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r242803456) [](ancestors = 242803456)

Reply:
They could be const though couldn't they?

---
In reply to: [243109370](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243109370) [](ancestors = 243109370,242803456)

Reply:
According to Xavier, we choose to build some structures during construction once because they are constant and we can re-use them during repeated execution of the instantiated operator so we have some savings. 
Regardless of this specific members, is it true or not?

---
In reply to: [243159429](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243159429) [](ancestors = 243159429,243145717,243109370,242803456)

Reply:
I can certainly make them const but since Compute() is const they are const anyway.

---
In reply to: [243145717](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243145717) [](ancestors = 243145717,243109370,242803456)